packages/core/useFetch/index.ts

```diff
@@ -276,60 +333,105 @@ export function useFetch<T>(url: MaybeRef<string>, ...args: any[]): UseFetchRetu
       }
     }
 
+    let isCanceled = false
+    const context: BeforeFetchContext = { url: unref(url), options: fetchOptions, cancel: () => { isCanceled = true } }
+
+    if (options.beforeFetch)
+      Object.assign(context, await options.beforeFetch(context))
+
+    if (isCanceled || !fetch)
+      return Promise.resolve()
+
     const defaultFetchOptions: RequestInit = {
       method: config.method,
       headers: {},
     }
 
-    if (config.payload) {
+    let payload = context.options?.body || config.payload
+
+    if (payload) {
```

Comment:
Most of the work below here does not need to be done. `fetch` and it's related types, like `Request`, will automatically figure out the proper `Content-Type` header.

![image](https://user-images.githubusercontent.com/1540597/121271331-1139af80-c889-11eb-8de5-b090c4bdd709.png)

It's worth noting that this is even the case with XHR. [I recently wrote some code related to this for XHR, which handles this](
![image](https://user-images.githubusercontent.com/1540597/121271447-4e9e3d00-c889-11eb-8fea-a5d03395d384.png)) that is much more concise. Hopefully that helps.

I _strongly_ advise that we don't do things like check a `payloadType` configuration or the like, because it will lead to user confusion and odd behaviors.

Reply:
We need to avoid to include a breaking change.
It is a pain `payloadType` option, we must deal with it.
The content type must be addressed since can be wrong, for example, adding it when using multipart, you can see the pictures included in this pr.

From a user perspective, it Will be more useful to provide the payload as a raw object rather than using fetch ones,.

I'm waiting to be included in a separated branch.